flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/stream/ExpireSnapshots.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.functions.async.AsyncRetryStrategy;
+import org.apache.flink.streaming.util.retryable.AsyncRetryStrategies;
+import org.apache.iceberg.flink.maintenance.operator.AsyncDeleteFiles;
+import org.apache.iceberg.flink.maintenance.operator.ExpireSnapshotsProcessor;
+import org.apache.iceberg.flink.maintenance.operator.TaskResult;
+import org.apache.iceberg.flink.maintenance.operator.Trigger;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class ExpireSnapshots {
+  private static final long DELETE_INITIAL_DELAY_MS = 10L;
+  private static final long DELETE_MAX_RETRY_DELAY_MS = 1000L;
+  private static final double DELETE_BACKOFF_MULTIPLIER = 1.5;
+  private static final long DELETE_TIMEOUT_MS = 10000L;
+  private static final int DELETE_PLANNING_WORKER_POOL_SIZE_DEFAULT = 10;
+  private static final int DELETE_ATTEMPT_NUM = 10;
+  private static final int DELETE_WORKER_POOL_SIZE_DEFAULT = 10;
+  private static final String EXECUTOR_TASK_NAME = "ES Executor";
+  @VisibleForTesting static final String DELETE_FILES_TASK_NAME = "Delete file";
+
+  private ExpireSnapshots() {
+    // Do not instantiate directly
+  }
+
+  /** Creates the builder for creating a stream which expires snapshots for the table. */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder extends MaintenanceTaskBuilder<ExpireSnapshots.Builder> {
+    private Duration minAge = null;
+    private Integer retainLast = null;
+    private int planningWorkerPoolSize = DELETE_PLANNING_WORKER_POOL_SIZE_DEFAULT;
+    private int deleteAttemptNum = DELETE_ATTEMPT_NUM;
+    private int deleteWorkerPoolSize = DELETE_WORKER_POOL_SIZE_DEFAULT;
+
+    /**
+     * The snapshots newer than this age will not be removed.
+     *
+     * @param newMinAge of the files to be removed
+     * @return for chained calls
+     */
+    public Builder minAge(Duration newMinAge) {
+      this.minAge = newMinAge;
+      return this;
+    }
+
+    /**
+     * The minimum {@link org.apache.iceberg.Snapshot}s to retain. For more details description see
+     * {@link org.apache.iceberg.ExpireSnapshots#retainLast(int)}.
+     *
+     * @param newRetainLast number of snapshots to retain
+     * @return for chained calls
+     */
+    public Builder retainLast(int newRetainLast) {
+      this.retainLast = newRetainLast;
+      return this;
+    }
+
+    /**
+     * The worker pool size used to calculate the files to delete.
+     *
+     * @param newPlanningWorkerPoolSize for planning files to delete
+     * @return for chained calls
+     */
+    public Builder planningWorkerPoolSize(int newPlanningWorkerPoolSize) {
```

Comment:
Are you suggesting to do everything in the operator instead of separating out the delete to another operator?

The Spark implementation even split the expired file calculations to multiple operators for performance reasons. In the long run we might go down that road... WYYT?

Reply:
> Are you suggesting to do everything in the operator instead of separating out the delete to another operator?

Nope, that is not what I meant.

I was wondering if `ExpireSnapshotsProcessor` and `AsyncDeleteFiles` should use the default shared thread pools from `ThreadPools.getWorkerPool()/getDeleteWorkerPool()`, instead of creating new pools.

Reply:
I don't like the idea of shared pools. If some users create multiple maintenance flows in a single job, they will end up using the same pool, and will block each-other.

Reply:
Also, if you have multiple slots on a single TaskManager, then these pools are shared between the subtasks. Which is again not something we want

Reply:
sharing thread pool is not necessarily a bad thing. it can limit the concurrent I/O. E.g., we may not want to have too many threads perform scan planing, which can be memory intensive.

Deletes have low memory footprint. Hence it is probably less of a concern to have separate pools. but probably good to keep an eye on the number of http connections

Reply:
We create our own pool for the source planner too, and it is not configurable there.
Do you think this risk is high enough to merit a new configuration value for this?

Reply:
for IcebergSource, there are two paths (1) batch: a separate pool is created and closed after the usage in the source constructor. this should be executed in the main thread of JM. so it has no concern. But there is a separate memory leak bug with incorrect usage of thread pools with shutdown hook that was discussed in the dev thread (2) streaming. the `ContinuousSplitPlannerImpl` code does support the shared pool although it is not actually used.

```
  public ContinuousSplitPlannerImpl(
      TableLoader tableLoader, ScanContext scanContext, String threadName) {
   ...
    this.isSharedPool = threadName == null;
    this.workerPool =
        isSharedPool
            ? ThreadPools.getWorkerPool()
            : ThreadPools.newWorkerPool(
                "iceberg-plan-worker-pool-" + threadName, scanContext.planParallelism());
  }
```

My main question is if we should use the shared pools (worker and delete) for maintenance tasks. 
```
  public static ExecutorService getWorkerPool() { 
   return WORKER_POOL;
  }
```

One option is that if `planningWorkerPoolSize` is not set, it is null and default to the shared pool.

Reply:
Moved to using the shared pool if the pool size is not set.